src/libs/ReportUtils.ts

```diff
@@ -5967,6 +5971,41 @@ function canLeavePolicyExpenseChat(report: OnyxEntry<Report>, policy: OnyxEntry<
     return isPolicyExpenseChat(report) && !(PolicyUtils.isPolicyAdmin(policy) || PolicyUtils.isPolicyOwner(policy, currentUserAccountID ?? -1) || isReportOwner(report));
 }
 
+/**
+ * Whether the user can join a report
+ */
+function canJoinChat(report: OnyxEntry<Report>, parentReportAction: OnyxEntry<ReportAction>, policy: OnyxEntry<Policy>): boolean {
+    if (ReportActionsUtils.isWhisperAction(parentReportAction)) {
+        return false;
+    }
+
+    if (report?.notificationPreference !== CONST.REPORT.NOTIFICATION_PREFERENCE.HIDDEN) {
+        return false;
+    }
+
+    if (isParentGroupChat(report)) {
+        return false;
+    }
+
+    return isUserCreatedPolicyRoom(report) || isChatReport(report) || canLeaveRoom(report, !isEmptyObject(policy)) || canLeavePolicyExpenseChat(report, policy);
```

Comment:
Why do we check for `isChatReport` here and not `isChatThread`?

Reply:
Did you explore if `canLeaveRoom` can be left out here? On the face of it, it seems like it shouldn't be here :smile:

Reply:
I agree. Let's get rid of all the `canLeaveWhatever` logic and just clearly document all the reasons we can leave or join (even if we have to repeat some stuff). I think that will be much less confusing.

Reply:
> Why do we check for isChatReport here and not isChatThread?

You're right, that is my mistake when I bring the logic from `HeaderView` into util.

Reply:
> I agree. Let's get rid of all the canLeaveWhatever logic and just clearly document all the reasons we can leave or join (even if we have to repeat some stuff). I think that will be much less confusing.

In fact, we will only be able to join a chat if we can leave it. If we have left the room, then its notification is hidden, the join option will appear, and vice versa. That is the reason we have the `canLeave...` condition here.

@jjcoffee @marcaaron What do you think if I update the `canJoinChat` like this

```
/**
 * Whether the user can join a report
 * - We can leave the chat
 * - The chat isn't a thread of a whisper action
 * - The user left the chat (the notification of its is hidden)
 */
function canJoinChat(report: OnyxEntry<Report>, parentReportAction: OnyxEntry<ReportAction>, policy: OnyxEntry<Policy>): boolean {
    if (ReportActionsUtils.isWhisperAction(parentReportAction)) {
        return false;
    }

    if (report?.notificationPreference !== CONST.REPORT.NOTIFICATION_PREFERENCE.HIDDEN) {
        return false;
    }

    return canLeaveChat(report, policy);

}
```

Reply:
Sorry, I guess I feel like a broken record now. But I just don't think that "joining" and "leaving" need to be related to each other. It's also a pretty weird concept. "if you can leave then you can join". It doesn't make sense 😄

Reply:
I would suggest renaming `canLeavePolicyExpenseChat` to `isNonAdminOrOwnerOfPolicyExpenseChat()`.

I also find this method a bit weird:

https://github.com/Expensify/App/blob/9a727de1e67013a8baf2dff018063af03d963ada/src/libs/ReportUtils.ts#L5261-L5278

It has a default of `true`. But aren't there specific types of reports we can leave? We should check if it's one of them. Instead we are checking if it's NOT one of the rooms we "can't" leave.

I bet once we figure out what those are then the logic is much simpler and we might not even need a `canLeaveRoom()` method at all.

Reply:
> I would suggest renaming canLeavePolicyExpenseChat to isNonAdminOrOwnerOfPolicyExpenseChat().

I checked this function and agree `isNonAdminOrOwnerOfPolicyExpenseChat` is good, it's what we do in this function.

Reply:
> I bet once we figure out what those are then the logic is much simpler and we might not even need a canLeaveRoom() method at all.

@marcaaron `canLeaveRoom` function just updated the case for invoice room, I think we can keep this logic as it is now. WDYT?

Reply:
@marcaaron [Here](https://github.com/Expensify/App/blob/573d9893ae2a8e6e4ebe64088cb2fbf62ce81804/src/libs/ReportUtils.ts#L5489-L5533)'s the latest `canLeaveRoom` for context. I'm starting to wonder if it's better to open a separate issue to tidy up the overall logic here :sweat_smile: 

That being said, the core issue here seems to be that 95% of `canLeaveRoom` is actually "is the room of a type that we can join or leave" with "is this a room we can leave" tacked on. Unless there _are_ cases where you can leave a room, but not join again? (Unlikely, but maybe I've missed something!)

Does it then make sense to clear things up a bit by creating a separate method, `canChangeRoomSubscription` or something, and then we just call that in both `canLeaveRoom` and `canJoinRoom` (along with `notificationPreference` checks)? (Though I realise this is almost going back to what we had before with `canLeaveOrJoinRoom` :smile:)

Reply:
I'd love if we could stick with the current plan and do the tidying here. More bugs and confusion will arise from this if we don't make an effort to improve what we've got.